### ISSUE TITLE
How to deal with tasks after an unexpected service interruption

### ISSUE THREAD

**Issue by Glf9832** [nameko/nameko#583]
I want to know that when the task in the service is still running, the service is interrupted, how to ensure that the task runs until it is completed. How did you do it? Thank you!

**Comment by derekjanni** [comment]
Here's a good test:

Here's our `helloworld.py` greeting service, slightly modified so it takes between 1-2 minutes to complete
```python
from nameko.rpc import rpc
import numpy as np
import time

class GreetingService:
	name = "greeting_service"

	@rpc
	def hello(self):
		start = time.time()
		time.sleep(np.random.randint(60,120))
		duration = time.time() - start
		return duration
```
We run the service with:
```
nameko run helloworld
```
And in a separate shell, lets invoke a call:
```python
from nameko.standalone.rpc import ClusterRpcProxy
import time

config = {
    'AMQP_URI': 'amqp://guest:guest@localhost:5672'  # out of the box Rabbit config. Change as you see fit
}

with ClusterRpcProxy(config) as cluster_rpc:
    print(cluster_rpc.greeting_service.hello())

```
Once that call is invoked, go to your [MQ management console](http://localhost:15672/#/queues/%2F/rpc-greeting_service) and look at it for a sec. Notice that there's a message that gets placed in the queue, but it will not disappear until your function completes while the service is running. Wait 30s after firing the rpc call above and then shut down your greeting service that was running with Ctrl+Z. You'll notice after leaving your calling application for 5 minutes that there's no response, even though the message should have been picked up and should have been processed. But since the service died, the consumer ack never happened - the task never reported back to MQ that it was complete and thus, it never really did complete from the client's perspective. I'm interested to try this with something more concrete like a file write or a db operation, but I don't have time for that right now. Might also be interesting to try with `call_async` and see what happens.

TL;DR - no guarantee that your task will complete if your service crashes. If you gently shut it down, I think the default behavior is to wait for the current active tasks to finish before shutting down, so you'll get a result from the call above, but if your service crashes you should expect the tasks running on it to crash as well

**Comment by mattbennett** [comment]
> TL;DR - no guarantee that your task will complete if your service crashes. If you gently shut it down, I think the default behavior is to wait for the current active tasks to finish before shutting down, so you'll get a result from the call above, but if your service crashes you should expect the tasks running on it to crash as well

This is correct. If your service crashes, the unack'd request message will be reclaimed by the broker when it detects that the consumer is gone. It's then available to be picked up by another consumer, and the RPC client will happily wait until it receives a response message from somewhere. There are currently no limits to the number of times that a message may be requeued.

The messaging guarantees are "at least once" delivery -- your request message will be processed at least once, maybe more than once. You should aim to make your RPC entrypoints idempotent because of this.

If you gracefully shut down a service, it will wait until it's finished processing all the current workers, and the request messages will be ack'd in that case.

There's no difference in behaviour between sync and async RPC except that the client doesn't wait for the result message.
